docs/specs/code/SIGNING.md

```diff
@@ -0,0 +1,183 @@
+
+# Proposal
+
+We propose a scheme that is generic in nature. By that we mean it does not specify a serialization format for the actual data you wish to sign nor does it enforce a specific communication channel between the signer and the validator.
+
+For the encoding, we will use [bech32](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) not only because it solves problems (1), (2) and (3) but because it will already be implemented for Cardano starting in Shelley as specified in [CIP-001](https://github.com/input-output-hk/implementation-decisions/blob/master/text/0001-address.md
+).
+
+### Brief note on CIP-001 on-chain encoding
+
+CIP-001 tries to minimize on-chain data by neither storing the checksum nor the protocol magic in its encoding. It's up to clients to display the correct bech32 prefix based on the context and not the on-chain data. Instead, it relies on the fact that the Cardano transaction signatures include the protocol magic inside them to avoid replay attacks. This is enough to protect end-users and, in order to protect developers, it also contains a single-bit discriminant that simply encodes testnet vs mainnet.
+
+### Brief note on transaction signing format
+
+Cardano uses [CBOR](http://cbor.io/) for most encoding. You can find a fairly good introduction for how Cardano uses CBOR for the legacy transaction format [here](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/tx/binary-format.md).
+
+All objects with special meaning when signed in Cardano are prefixed by a CBOR tag. For example, transaction signatures are generated by signing the following
+
+```rust
+se
+  .write_unsigned_integer(SigningTag::Tx as u64)
+  .serialize(&protocol_magic)
+  .serialize(txid)
+```
+
+and these are wrapped in a `witness` (serialization logic [here](https://github.com/input-output-hk/rust-cardano/blob/103d15d2df17f6ffc055f4563da16c206e7ce135/cardano/src/tx.rs#L202))
+
+lastly, an array of witnesses is represented by a fixed-length CBOR array (call this function `FIXED_ARRAY`)
+
+```rust
+serializer.write_array(Len::Len(data.len() as u64))?;
+for element in data {
+    Serialize::serialize(element, serializer)?;
+}
+```
+
+You can see all existing tags in Rust [here](https://github.com/input-output-hk/rust-cardano/blob/2e924a4457203fa6b19f409cb897d90cd6b699f6/cardano/src/tags.rs) and in Haskell [here](https://github.com/input-output-hk/cardano-sl/blob/ccca8b346413f12e9fad8cab4c855e69669a2151/crypto/Pos/Crypto/Signing/Tag.hs#L15)
+
+### Proposed format
+
+##### Definitions
+
+**Define**: New CBOR tag `MessageTag = 0x0b` to be used for our new message signing standard. This tag is different than all other tags for meaningful signed data inside Cardano.
+
+**Define**: byte array `UserMessage` which is the message the user wishes to sign. It is RECOMMENDED to use unicode with utf8 encoding if the message is a string and RECOMMENDED to use network-byte-order for numbers.
```

Comment:
Now, this begs a question
1) what should wallets do (from the UX perspective) if the content is binary data?
2) hardware wallets are not likely to support utf8. We might consider recommending ascii instead
3) we might consider going further and have at least a some idea of what the ascii format might look like (a very optional recommendation)

Reply:
Yeah I tried to look for examples of how other blockchains have tackled this but I can find surprisingly little information about this. It seems like the most common solution is that the wallet UI shows you a hash of the message you're about to sign and your hardware wallet shows this hash on the display (which ensures it's ASCII). It means that you have to trust the wallet to hash the message honestly which is unfortunate.

We can do something different like 
If ASCII, display the message as-is
If non-ASCII, use the hash method

then we add to this spec that we RECOMMEND using ASCII